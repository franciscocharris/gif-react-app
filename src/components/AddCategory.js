import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');
    const handleInputChange = (e) =>{
        setInputValue( e.target.value  );
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length > 0){
            // console.log(`handleSubmit`, inputValue);
            // al llamar a una funcion de un hook, traera los valores si se trata como un callback
            // aqui se esta llenando el array de los gifs que se van a cargar
            setCategories( cat => [inputValue, ...cat] ); 
            setInputValue('');  
        }
    }

    return (
        <form onSubmit={handleSubmit} >
            {/* solo prueba */}
            <p>{ inputValue }</p>
            <input 
                type='text'
                placeholder='Buscar Gifs'
                value={inputValue}
                // el onchange se dispara cada vez que cambien el valor del input
                onChange={ handleInputChange }
            />
        </form>
        
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}