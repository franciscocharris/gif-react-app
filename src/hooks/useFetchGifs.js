import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

// l adiferencia de un custom hook de un functional component, es que ambos son funciones, pero el hook
// no nesesariamemnte retorna jsx
export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGifs(category)
        // una vez la promesa resuelva las imagenes, se devuelve o cambia el estado dando los nuevos valores
        // y se retorna, para que pueda ser renderizados
            .then( imgs => {

                // console.log(imgs);
                setState({
                    data: imgs,
                    loading: false
                });
            });
    }, [category]); //si la categoria cambia entonces el useEffect debe volver a ejecutarse, parea traer los gifs

    return state; //devuelve un objeto con un estado inicial que es un objeto
}