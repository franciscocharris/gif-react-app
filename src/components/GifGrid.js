import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

// para renombrar con la desestructuracion, solo es dos puntos : y el nuevo nombre
    const { data:images, loading} = useFetchGifs(category);
    
    // el useEffect sirve para ejecutar codigo con una condicion, se le pasa parametros y si es un array vacio
    // entonces significa que se ejecute una sola vez, cuamdo se renderize el componente
    // useEffect( () => {
    //     getGifs( category )
    //         // .then( imgs => setImages(imgs) );
    //         .then( setImages);
    // }, [category]);

    return (
        <>
            <h3 className='animate__animated animate__animated' ><span>{ category }</span></h3>
            {loading && <p className='animate__animated animate__flash'>Loading</p> }
            <div className='card-grid'>
                    {
                        images.map( img => 
                            <GifGridItem 
                                key={img.id} 
                                // se le hace un spread operator a la coleccion de objetos para enviar propiedades de manera
                                // individual
                                {...img}
                            />
                        )
                    }
            </div>
        </>
    )
}
