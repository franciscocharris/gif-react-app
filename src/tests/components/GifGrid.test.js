import { shallow } from 'enzyme';
import React from 'react';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs');

describe('pruebas en el componente de GifGrid', () => {
    
    const category = 'naruto';
    test('debe coincidir la renderizacion con el snapshot', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={ category } />);


        expect(wrapper).toMatchSnapshot();
    });
    
    // test('debe mostrar los items correctamnete', () => {

    //     // con el mock uno da o simula los valores, en este caso esto es lo que retorna por defeto

    //     const gifs = [{
    //         id: 'abc',
    //         title: 'hola',
    //         url: 'https://mega.nz/fm/5IMy3boB'
    //     }];
    //     // con esto estamos simulando como si el hook useFetchGifs estuviera rettonando esos valores, y despues 
    //     // el componente lo renderiza ciomo normalmente lo haria
    //     useFetchGifs.mockReturnValue({
    //         data: gifs,
    //         loading: false
    //     });
    //     const wrapper = shallow(<GifGrid category={ category } />);

    //     expect(wrapper).toMatchSnapshot();
        
    // })

    // mas preciso que el de arriba
    test('debe mostrar los items correctamnete ', () => {
        const gifs = [{
            id: 'abc',
            title: 'hola',
            url: 'https://mega.nz/fm/5IMy3boB'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false 
        });

        const wrapper = shallow(<GifGrid category={ category } />);
        // expect(wrapper).toMatchSnapshot();
        // primero la p del loading no deberia estar
        expect(wrapper.find('p').exists()).toBe(false);

        // la longitud de componentes renderizados desbe ser igual al numero del objeto
        expect(wrapper.find('GifGridItem').length).toBe( gifs.length);
    })
    
    
    
})
