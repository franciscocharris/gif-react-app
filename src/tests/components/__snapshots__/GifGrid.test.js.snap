// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el componente de GifGrid debe coincidir la renderizacion con el snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__animated"
  >
    <span>
      naruto
    </span>
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
